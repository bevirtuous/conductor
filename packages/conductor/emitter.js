module.exports = require('./src/emitter');
