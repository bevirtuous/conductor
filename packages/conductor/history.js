module.exports = require('./src/history');
