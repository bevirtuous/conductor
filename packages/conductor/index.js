module.exports = require('./src/Conductor');
