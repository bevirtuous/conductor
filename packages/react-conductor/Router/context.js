module.exports = require('../src/context');