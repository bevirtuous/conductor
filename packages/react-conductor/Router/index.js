const { RouteContext } = require('../src/context');
module.exports = require('../src/Router');
module.exports.RouteContext = RouteContext;