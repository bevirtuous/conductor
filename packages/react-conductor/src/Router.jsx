import React, { Component } from 'react';
import PropTypes from 'prop-types';
import conductor from '@virtuous/conductor';
import * as events from '@virtuous/conductor-events';
import getRouteStack from '@virtuous/conductor-helpers/src/getRouteStack';

/**
 * The Router component.
 */
class Router extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    initialPath: PropTypes.string.isRequired,
    initialPattern: PropTypes.string.isRequired,
  };

  /**
   * The constructor.
   * @param {Object} props The component props.
   */
  constructor(props) {
    super(props);

    this.state = {
      routeStack: [{
        path: props.initialPath,
        pattern: props.initialPattern,
      }],
    };

    // Setup some listeners for router events.
    events.onPush(this.handleRouteChange);
    events.onPop(this.handleRouteChange);
    events.onReplace(this.handleRouteChange);

    // Push the first route.
    setTimeout(() => {
      conductor.push(props.initialPath);
    }, 0);
  }

  /**
   * Limit re-rendering unless the state actually changes.
   *
   * @param {Object} nextProps The next props.
   * @param {Object} nextState The next state.
   * @returns {boolean}
   */
  shouldComponentUpdate() {
    // TODO: Add some sort of check.
    return true;
  }

  /**
   * Find the index of the given route pattern against the currently open routes.
   * @param {string} pattern The pattern to check for.
   * @returns {number} The index of the pattern.
   */
  getRouteIndex(pattern) {
    for (let i = 0, j = this.state.routeStack.length; i < j; i += 1) {
      if (this.state.routeStack[i].pattern === pattern) {
        return i;
      }
    }
    return -1;
  }

  /**
   * Update the state based on the router changes.
   * @param {string} action The history action performed.
   * @param {string} currentPath The newly opened path.
   * @param {string} prevPath The previous path.
   * @param {string} routeStack The open routes.
   */
  handleRouteChange = () => {
    const routeStack = getRouteStack();
    this.setState({ routeStack });
  }

  /**
   * The render method.
   * @returns {JSX}
   */
  render() {
    /**
     * We need to apply some props to the children based on
     * the current router state.
     */
    // TODO: Move this code to another function.
    return React.Children.map(this.props.children, (route, index) => {

      // TODO: This is not the correct way to get the index.
      // Get the index of this route.
      const routeIndex = this.getRouteIndex(route.props.pattern);

      // Check if this route is open.
      const isOpen = routeIndex > -1;

      // Check if this route is the current visible route.
      const isVisible = routeIndex === this.state.routeStack.length - 1;

      // If this route is open then pass the pathname as a prop.
      const path = isOpen ? this.state.routeStack[routeIndex].pathname : null;
      const state = isOpen ? this.state.routeStack[routeIndex].state : null;

      return React.cloneElement(route, {
        index,
        isOpen,
        isVisible,
        path,
        state,
      });
    });
  }
}

export default Router;
