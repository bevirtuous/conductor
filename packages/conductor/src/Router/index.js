import uuid from 'uuid/v4';
import Route from '../Route';
import * as errors from './errors';
import emitter from '../emitter';
import history from '../history';
import matcher from '../matcher';
import stack from '../Stack';
import * as constants from '../constants';

/**
 * The Router class.
 */
class Router {
  /**
   * TODO:
   */
  constructor(createHistory = history) {
    // Flag to indicate a native history event. Should always be reset to true.
    this.nativeEvent = true;

    this.history = createHistory();

    // The patterns are collected to match against pathnames.
    this.patterns = {};

    // Flag to indicate whether or not there is an ongoing history change.
    this.routing = false;

    // The `routeIndex` is used to track which stack entry is the current route.
    this.routeIndex = 0;

    this.action = constants.ACTION_PUSH;

    // Unsubscribe to any other history module changes.
    if (typeof this.historyListener === 'function') {
      this.historyListener();
      stack.clear();
    }

    // 
    this.addInitialRoute();

    this.historyListener = this.history.listen(this.handleNativeEvent);
  }

  /**
   * 
   */
  handleNativeEvent = (location, action) => {
    if (!this.nativeEvent) {
      return;
    }

    const next = stack.getByIndex(this.routeIndex + 1);

    if (next && location.state && location.state.route && next.id === location.state.route.id) {
      this.handlePush({
        pathname: location.pathname,
        state: location.state,
      }, false);

      return;
    }

    if (action === constants.ACTION_POP) {
      this.handlePop();
    }
  }

  /**
   * Populate the stack with an initial entry to match the history module.
   * Note: we cannot match it against a pattern at this point.
   */
  addInitialRoute = () => {
    const { hash, pathname, search } = this.history.location;
    const fullPathname = `${pathname}${search}${hash}`;
    const id = uuid();
    stack.add(id, new Route({ id, pathname: fullPathname }));
  }

  /**
   * @param {Object} params The router action params.
   * @returns {Promise}
   */
  handlePop = (params = {}) => {
    return new Promise((resolve, reject) => {
      const {
        emitBefore = true,
        emitAfter = true,
        steps = 1,
        state = null,
      } = params;
      let unlisten = null;
      const { size } = stack.getAll();

      if (size < 2) {
        reject(new Error(errors.ESTACKLENGTH));
        this.nativeEvent = true;
        return;
      }

      if (steps <= 0) {
        reject(new Error(errors.EINVALIDSTEPS));
        this.nativeEvent = true;
        return;
      }

      // Get id of target route.
      const targetIndex = Math.max(this.routeIndex - steps, 0);
      const prev = stack.getByIndex(this.routeIndex);
      const next = stack.getByIndex(targetIndex);
      const end = { prev, next };

      // Emit creation event.
      if (emitBefore) {
        emitter.emit(constants.EVENT_WILL_POP, end);
      }

      if (state) {
        next.state = Object.assign(next.state, state);
      }

      /**
       * 
       */
      const callback = () => {
        unlisten();
        this.routeIndex = targetIndex;
        this.action = constants.ACTION_POP;

        if (emitAfter) {
          emitter.emit(constants.EVENT_DID_POP, end);
        }

        resolve(end);
        this.nativeEvent = true;
        this.routing = false;
      };

      /**
       * Create a reference to the history listener
       * to be able to unsubscribe from inside the callback.
       */
      unlisten = this.history.listen(callback);

      // Perform the history back action.
      if (!this.nativeEvent) {
        this.history.go(steps * -1);
      } else {
        callback();
      }
    });
  }

  /**
   * @param {Object} params The params to use when navigating.
   * @returns {Promise}
   */
  handlePush(params, override = true) {
    return new Promise((resolve, reject) => {
      // Check for missing parameters.
      if (!params) {
        reject(new Error(errors.EPARAMSMISSING));
        this.nativeEvent = true;
        return;
      }

      // Check for empty params.
      if (Object.keys(params).length === 0) {
        reject(new Error(errors.EPARAMSEMPTY));
        this.nativeEvent = true;
        return;
      }

      // Check for ongoing router action.
      if (this.routing) {
        // TODO: create real error
        reject(new Error('Error'));
        this.nativeEvent = true;
        return;
      }

      const {
        emitBefore = true,
        emitAfter = true,
        pathname,
        state,
      } = params;
      const pattern = this.findPattern(pathname.split('?')[0]);
      let unlisten = null;

      if (!pattern) {
        reject(new Error(errors.EINVALIDPATHNAME));
        this.nativeEvent = true;
        return;
      }

      // Block further router actions until this Promise has been returned.
      this.routing = true;

      // Remove all unwanted items from the stack.
      if (override) {
        while (this.routeIndex < stack.getAll().size - 1) {
          const [id] = stack.last();
          stack.remove(id);
        }
      }

      const id = uuid();
      const { transform } = this.patterns[pattern];
      const prev = stack.getByIndex(this.routeIndex);
      const next = new Route({
        id,
        pathname,
        pattern,
        state,
        transform,
      });

      // Add item to the stack
      stack.add(id, next);

      // Emit creation event.
      if (emitBefore) {
        emitter.emit(constants.EVENT_WILL_PUSH, { prev, next });
      }

      /**
       * TODO: move to class function with params
       * The history event callback.
       */
      const callback = () => {
        // Unsubscribe from the history events.
        unlisten();

        // Increment the route index.
        this.routeIndex += 1;

        this.action = constants.ACTION_PUSH;

        // Emit completion event.
        if (emitAfter) {
          emitter.emit(constants.EVENT_DID_PUSH, { prev, next });
        }

        // Resolve the Promise.
        resolve({ prev, next });
        this.nativeEvent = true;
        this.routing = false;
      };

      /**
       * Create a reference to the history listener
       * to be able to unsubscribe from inside the callback.
       */
      unlisten = this.history.listen(callback);

      // Perform the history push action.
      if (!this.nativeEvent) {
        this.history.push({
          pathname,
          state: {
            ...state,
            route: { id },
          },
        });
      } else {
        callback();
      }
    });
  }

  /**
   * Match the given pathname to a registered pattern.
   * @param {string} pathname The pathname to match.
   * @returns {string|null}
   */
  findPattern = (pathname) => {
    const pattern = Object.keys(this.patterns).find(key => this.patterns[key].match(pathname));
    return pattern || null;
  }

  /**
   * Registers a route pattern to match new pathnames against.
   * @param {string} pattern The pattern to register.
   * @param {Function} transform The route transformer.
   */
  // TODO: convert to Promise
  register = (pattern, transform = null) => {
    if (!pattern) {
      throw new Error(errors.EMISSINGPATTERN);
    }

    if (typeof pattern !== 'string') {
      throw new Error(errors.EINVALIDPATTERN);
    }

    // 
    const match = matcher(pattern);

    // 
    this.patterns[pattern] = {
      match,
      transform,
    };

    // Find the pathname of the first route.
    const [, route] = stack.first();

    // If it has been set then we don't need to match it.
    if (route.pattern !== null) {
      return;
    }

    // 
    if (match(route.pathname)) {
      route.setPattern(pattern);

      const end = { prev: null, next: route };

      emitter.emit(constants.EVENT_WILL_PUSH, end);
      emitter.emit(constants.EVENT_DID_PUSH, end);
    }
  }

  handleReplace = (params) => {
    return new Promise((resolve, reject) => {
      // Check for missing parameters.
      if (!params) {
        reject(new Error(errors.EPARAMSMISSING));
        this.nativeEvent = true;
        return;
      }

      // Check for empty params.
      if (Object.keys(params).length === 0) {
        reject(new Error(errors.EPARAMSEMPTY));
        this.nativeEvent = true;
        return;
      }

      // Check for ongoing router action.
      if (this.routing) {
        // TODO: create real error
        reject(new Error('Error'));
        this.nativeEvent = true;
        return;
      }

      const {
        emitBefore = true,
        emitAfter = true,
        pathname,
        state,
      } = params;
      const pattern = this.findPattern(pathname.split('?')[0]);
      let unlisten = null;

      if (!pattern) {
        reject(new Error(errors.EINVALIDPATHNAME));
        this.nativeEvent = true;
        return;
      }

      // Block further router actions until this Promise has been returned.
      this.routing = true;

      const { id } = stack.getByIndex(this.routeIndex);
      const { transform } = this.patterns[pattern];
      const prev = stack.get(id);
      const next = new Route({
        id,
        pathname,
        pattern,
        state,
        transform,
      });
      const end = { prev, next };

      // Add item to the stack
      stack.add(id, next);

      // Emit creation event.
      if (emitBefore) {
        emitter.emit(constants.EVENT_WILL_REPLACE, end);
      }

      /**
       * The history event callback.
       */
      const callback = () => {
        // Unsubscribe from the history events.
        unlisten();

        this.action = constants.ACTION_REPLACE;

        // Emit completion event.
        if (emitAfter) {
          emitter.emit(constants.EVENT_DID_REPLACE, end);
        }

        resolve(end);
        this.nativeEvent = true;
        this.routing = false;
      };

      /**
       * Create a reference to the history listener
       * to be able to unsubscribe from inside the callback.
       */
      unlisten = this.history.listen(callback);

      // Perform the history replace action.
      if (!this.nativeEvent) {
        this.history.replace({
          pathname,
          state: {
            ...state,
            route: { id },
          },
        });
      } else {
        callback();
      }
    });
  }

  /**
   * @param {Object} params The params when routing.
   * @returns {Promise}
   */
  push = (params) => {
    this.nativeEvent = false;
    return this.handlePush(params);
  }

  /**
   * @param {Object} params The params when routing.
   * @returns {Promise}
   */
  pop = (params) => {
    this.nativeEvent = false;
    return this.handlePop(params);
  }

  /**
   * @param {Object} params The params when routing.
   * @returns {Promise}
   */
  replace = (params) => {
    this.nativeEvent = false;
    return this.handleReplace(params);
  }

  /**
   * 
   */
  reset = () => {
    return new Promise((resolve, reject) => {
      const { size } = stack.getAll();
      const [, route] = stack.first();

      const prev = stack.getByIndex(this.routeIndex);
      const end = {
        prev,
        next: route,
      };

      if (size === 1) {
        // TODO: Add error
        reject();
        return;
      }

      emitter.emit(constants.EVENT_WILL_RESET, end);

      this.pop({
        emitBefore: false,
        emitAfter: false,
        steps: size,
      }).then(() => {
        emitter.emit(constants.EVENT_DID_RESET, end);
        resolve(end);
      });
    });
  }

  /**
   * @param {string} pathname The pathname to reset to.
   * @returns {Promise}
   */
  resetTo = (pathname) => {
    return new Promise((resolve, reject) => {
      if (!pathname) {
        reject(new Error(errors.EMISSINGPATHNAME));
        return;
      }

      if (!this.findPattern(pathname)) {
        reject(new Error(errors.EINVALIDPATHNAME));
        return;
      }

      const id = uuid();
      const prev = stack.getByIndex(this.routeIndex);
      const next = new Route({ id, pathname });
      const end = { prev, next };

      emitter.emit(constants.EVENT_WILL_RESET, end);
      stack.reset([id, next]);
      emitter.emit(constants.EVENT_DID_RESET, end);

      resolve(end);
    });
  }

  /**
   * 
   */
  update = (id, state = {}, emit = true) => {
    return new Promise((resolve, reject) => {
      //
      if (!id || Object.keys(state).length === 0) {
        reject(new Error(errors.EPARAMSINVALID));
        return;
      }

      const route = stack.get(id);

      if (!route) {
        reject(new Error(errors.EINVALIDID));
        return;
      }

      route.state = {
        ...route.state,
        ...state,
      };
      route.updated = Date.now();

      stack.update(id, route);

      if (emit) {
        emitter.emit(constants.EVENT_UPDATE, route);
      }

      resolve(route);
    });
  }

  // TODO: Add deprecation warning
  getCurrentRoute = () => (
    stack.getByIndex(this.routeIndex)
  )
}

export default new Router();
