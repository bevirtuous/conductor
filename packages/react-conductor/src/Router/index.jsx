import React from 'react';
import PropTypes from 'prop-types';
import {
  router,
  stack as routeStack,
  onDidPush,
  onDidPop,
  onDidReplace,
  onDidReset,
} from '@virtuous/conductor';
import { RouterContext } from '../context';

/**
 * The Router component.
 */
class Router extends React.Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    history: PropTypes.func,
  }

  static defaultProps = {
    history: null,
  }

  /**
   * @param {Object} props The component props.
   */
  constructor(props) {
    super(props);

    if (typeof props.history === 'function') {
      router.constructor(props.history);
    }

    this.state = {
      stack: Array.from(routeStack.getAll()),
      updated: null,
    };

    onDidPush(this.update);
    onDidPop(this.update);
    onDidReplace(this.update);
    onDidReset(this.update);
  }

  /**
   * @param {Object} nextProps The next component props.
   * @param {Object} nextState The next component state.
   * @returns {boolean}
   */
  shouldComponentUpdate(nextProps, nextState) {
    const { updated } = this.state;
    return updated !== nextState.updated;
  }

  /**
   * Update the local stack with the route stack.
   */
  update = () => {
    this.setState({
      stack: Array.from(routeStack.getAll()),
      updated: Date.now(),
    });
  }

  /**
   * @returns {JSX}
   */
  render() {
    const { children } = this.props;
    const { stack } = this.state;

    return (
      <RouterContext.Provider value={stack}>
        {children}
      </RouterContext.Provider>
    );
  }
}

export default Router;
