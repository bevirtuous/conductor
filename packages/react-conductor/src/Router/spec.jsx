import React from 'react';
import { shallow, mount } from 'enzyme';
import { router } from '@virtuous/conductor';
import Router from './index';

const spy = jest.spyOn(router, 'constructor');

describe('<Router />', () => {
  it('should render with children and context', () => {
    const app = mount((
      <Router>
        <p />
      </Router>
    ));

    const child = app.find('p');
    expect(child.length).toBe(1);
    expect(app).toMatchSnapshot();

    // TODO: Test Router context value
  });

  it('should re-render with children and context', () => {
    // TODO: Test Router context value
  });

  it.only('should pass the history function to core', () => {
    const mockHistory = () => ({
      listen: () => {},
      location: {},
    });

    shallow((
      <Router history={mockHistory}>
        <p />
      </Router>
    ));

    expect(spy).toHaveBeenCalledWith(mockHistory);
  });
});
