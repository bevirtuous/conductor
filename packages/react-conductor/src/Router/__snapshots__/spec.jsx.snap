// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Router /> should render with children and context 1`] = `
<Router
  history={null}
>
  <p />
</Router>
`;
